src/v0/destinations/tiktok_ads/transformV2.js

```diff
@@ -0,0 +1,741 @@
+/* eslint-disable camelcase */
+/* eslint-disable @typescript-eslint/naming-convention */
+const set = require('set-value');
+const { ConfigurationError, InstrumentationError } = require('@rudderstack/integrations-lib');
+const { EventType } = require('../../../constants');
+const {
+  constructPayload,
+  defaultRequestConfig,
+  defaultPostRequestConfig,
+  removeUndefinedAndNullValues,
+  defaultBatchRequestConfig,
+  getSuccessRespEvents,
+  isDefinedAndNotNullAndNotEmpty,
+  getDestinationExternalID,
+  getHashFromArrayWithDuplicate,
+  checkInvalidRtTfEvents,
+  handleRtTfSingleEventError,
+} = require('../../util');
+const { getContents, hashUserField } = require('./util');
+const {
+  trackMappingV2,
+  trackEndpointV2,
+  eventNameMapping,
+  PARTNER_NAME,
+  maxBatchSizeV2,
+} = require('./config');
+const { JSON_MIME_TYPE } = require('../../util/constant');
+
+/**
+ * Generated track payload for an event using TikTokTrackV2.json config mapping,
+ * hashing user properties and
+ * defining contents from products
+ * @param {*} message
+ * @param {*} Config
+ * @param {*} event
+ * @returns track payload
+ */
+const getTrackResponsePayload = (message, Config, event) => {
+  const payload = constructPayload(message, trackMappingV2);
+
+  // if contents is not an array converting it into array
+  if (payload.properties?.contents && !Array.isArray(payload.properties.contents)) {
+    payload.properties.contents = [payload.properties.contents];
+  }
+
+  // if contents is not present but we have properties.products present which has fields with superset of contents fields
+  if (payload.properties && !payload.properties?.contents && message.properties?.products) {
```

Comment:
you can omit "?" in !payload.properties?.contents

Reply:
Sometimes the guys don't send properties resulting in bugsnag alert

Reply:
but it is already getting checked "payload.properties"

Reply:
removed